source/common/router/router.cc

```diff
@@ -851,26 +882,36 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
                    (route_entry_ && route_entry_->internalRedirectPolicy().enabled());
   if (buffering &&
       getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) {
-    ENVOY_LOG(debug,
-              "The request payload has at least {} bytes data which exceeds buffer limit {}. Give "
-              "up on the retry/shadow.",
-              getLength(callbacks_->decodingBuffer()) + data.length(), retry_shadow_buffer_limit_);
-    cluster_->trafficStats()->retry_or_shadow_abandoned_.inc();
-    retry_state_.reset();
-    buffering = false;
-    active_shadow_policies_.clear();
-    request_buffer_overflowed_ = true;
-
-    // If we had to abandon buffering and there's no request in progress, abort the request and
-    // clean up. This happens if the initial upstream request failed, and we are currently waiting
-    // for a backoff timer before starting the next upstream attempt.
-    if (upstream_requests_.empty()) {
-      cleanup();
-      callbacks_->sendLocalReply(
-          Http::Code::InsufficientStorage, "exceeded request buffer limit while retrying upstream",
-          modify_headers_, absl::nullopt,
-          StreamInfo::ResponseCodeDetails::get().RequestPayloadExceededRetryBufferLimit);
-      return Http::FilterDataStatus::StopIterationNoBuffer;
+    // Currently waiting to get a connection with the upstream. Increase the buffer limit
+    // accordingly.
+    if (wait_for_connect_ && getLength(callbacks_->decodingBuffer()) + data.length() <
+                                 retry_connection_failure_buffer_limit_) {
+      // Increasing buffer limit if necessary.
+      if (callbacks_->decoderBufferLimit() <=
+          getLength(callbacks_->decodingBuffer()) + data.length()) {
+        callbacks_->setDecoderBufferLimit(retry_connection_failure_buffer_limit_);
```

Comment:
this looks like it does a bunch of work each time we're over the limit.  Should we check the limit to only bump it up once?

Reply:
I looks like from the latest test that we can not call setDecoderBufferLimit in that case where we are using `Http::AsyncStreamImpl`.

I therefor increase the buffer limit by default in the decodeHeader.
see failed test: https://github.com/envoyproxy/envoy/actions/runs/8538790525/job/23392163514